ledger/lruonlineaccts.go

```diff
@@ -0,0 +1,121 @@
+// Copyright (C) 2019-2022 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package ledger
+
+import (
+	"github.com/algorand/go-algorand/data/basics"
+	"github.com/algorand/go-algorand/logging"
+)
+
+// lruAccounts provides a storage class for the most recently used accounts data.
+// It doesn't have any synchronization primitive on it's own and require to be
+// synchronized by the caller.
+type lruOnlineAccounts struct {
+	// accountsList contain the list of persistedAccountData, where the front ones are the most "fresh"
+	// and the ones on the back are the oldest.
+	accountsList *persistedOnlineAccountDataList
+	// accounts provides fast access to the various elements in the list by using the account address
+	accounts map[basics.Address]*persistedOnlineAccountDataListNode
+	// pendingAccounts are used as a way to avoid taking a write-lock. When the caller needs to "materialize" these,
+	// it would call flushPendingWrites and these would be merged into the accounts/accountsList
+	pendingAccounts chan persistedOnlineAccountData
+	// log interface; used for logging the threshold event.
+	log logging.Logger
+	// pendingWritesWarnThreshold is the threshold beyond we would write a warning for exceeding the number of pendingAccounts entries
+	pendingWritesWarnThreshold int
+}
+
+// init initializes the lruAccounts for use.
+// thread locking semantics : write lock
+func (m *lruOnlineAccounts) init(log logging.Logger, pendingWrites int, pendingWritesWarnThreshold int) {
+	m.accountsList = newPersistedOnlineAccountList().allocateFreeNodes(pendingWrites)
+	m.accounts = make(map[basics.Address]*persistedOnlineAccountDataListNode, pendingWrites)
+	m.pendingAccounts = make(chan persistedOnlineAccountData, pendingWrites)
+	m.log = log
+	m.pendingWritesWarnThreshold = pendingWritesWarnThreshold
+}
+
+// read the persistedAccountData object that the lruAccounts has for the given address.
+// thread locking semantics : read lock
+func (m *lruOnlineAccounts) read(addr basics.Address) (data persistedOnlineAccountData, has bool) {
```

Comment:
Why doesn't this refresh the value by moving it to the front?

Reply:
the caller does that manually, at its own discretion

Reply:
We want to make these LRU caches more generic, rather than have so duplication, would be a good feature to add then